include/layers/FCLayer.hpp

```diff
@@ -2,27 +2,78 @@
 #include <stdexcept>
 #include <vector>
 
+template <typename T>
+class Shape2D {
+ public:
+  Shape2D() : data_(), width_(), height_() {}
+  Shape2D(size_t height, size_t width = 1) : data_(width * height) {
+    this->width_ = width;
+    this->height_ = height;
+  }
+  Shape2D(size_t height, size_t width, const std::vector<T>& data)
+      : data_(data) {
+    this->width_ = width;
+    this->height_ = height;
+    data_.resize(width_ * height_, T(0));
+  }
+  Shape2D(const std::vector<T>& data)
+      : data_(data), width_(1), height_(data.size()) {}
+  Shape2D(const Shape2D& c) : data_(c.data_) {
+    this->width_ = c.width_;
+    this->height_ = c.height_;
+  }
+  Shape2D& operator=(const Shape2D& c) {
+    this->data_ = c.data_;
+    this->width_ = c.width_;
+    this->height_ = c.height_;
+    return *this;
+  }
+  T get(size_t i, size_t j = 0) const {
+    if (i >= height_ || j >= width_) {
+      throw std::out_of_range("Bad shape index");
+    }
+    return data_[i * width_ + j];
+  }
+  void set(size_t i, size_t j, const T& value) {
+    if (i >= height_ || j >= width_) {
+      throw std::out_of_range("Bad shape index");
+    }
+    data_[i * width_ + j] = value;
+  }
+  size_t get_width() const { return width_; }
+  size_t get_height() const { return height_; }
+  size_t size() const { return width_ * height_; }
+  void resize(size_t height, size_t width) {
+    this->width_ = width;
+    this->height_ = height;
+    data_.resize(width_ * height_, T(0));
+  }
+
+ private:
+  std::vector<T> data_;
```

Comment:
Shapes are usually not connected to the data. They just define dimensions.

Reply:
If they aren't connected, abstract Layer's run() function will now return std::pair(vector<ValueType> data, Shape shape) ?

Reply:
Shape is actually a part of the layer description

Reply:
Meaning that we have input shape, output shape

Reply:
Changed